TASK 1

Since the program's inception in 2009, dublinbikes bicycles have been used on more than 27.6 million trips. More than 14% of all trips in the city center are now made by cyclists, many of whom use the dublinbikes cycling network. In particular, the program has 67,136 yearly subscribers, with 96% of them using free trips in 2019. Since its launch, Dublin City Council has benefited from dublinbikes' assistance in promoting a modal shift away from short-distance driving, which has led to less traffic congestion, fewer greenhouse gas emissions, and improvements to population health and the environment. 

The data set was obtained from our api hosted by JCDecaux and is a real-time bike data set with 13 columns and 1130 attributes per column for weekdays and weekends.(http://www.dublinbikes.ie/Magazine/Reports/Just-Eat-dublinbikes-latestfigures
[Retrieved on 07/19/2019])

In this research, the availability of bicycles and bike stands on weekdays and weekends is examined in greater detail. The following inquiries are designed to help us fully comprehend how time affects Dublin's availability of bikes and bike stands. The following question would be investigated and a result reached using the Dublin Bike data set.
1. What is the average availabe bike on weekdays compared to weekends
2. What is the busiest & quietest station during weekdays and weekends
3. What is the number of available bikes for the answer of question 2(busiest & quietest station) for weekdays and weekends
4. Does the presence of a credit card enabled facility influence bike useage on weekdays and weekends
5. What is the total available bike stand, bike stand and available bikes that has a credit card enabled facility in a working day in contrast to a bank holiday?

By providing answers to the aforementioned questions, we may better understand the differences between weekday and weekend bike usage in Dublin, as well as some of the factors that affect it, and draw conclusions.

In [2]:
import pandas as pd
import numpy as np
from pandas import DataFrame
import warnings
warnings.filterwarnings("ignore")
import requests
import json

wkday = pd.read_csv("weekday_dublin_bike.csv")

A dateframe is created and the columns are renamed to suit the exploratory analysis. Afterwards, the columns that will be used are selected

In [4]:

wkday = wkday.rename(columns = {"NUMBER": "NUMBER", "CONTRACT NAME":"CONTRACT_NAME", "name":"NAME", "address":"ADDRESS","position":"POSITION","BANKING":"BANKING",
       "bonus":"BONUS","latitude":"LATITUDE", "longitude":"LONGITUDE","BIKE STANDS":"BIKE_STANDS","AVAILABLE BIKE STANDS":"AVAILABLE_BIKE_STANDS","AVAILABLE BIKES":"AVAILABLE_BIKES",
       "status":"STATUS","LAST UPDATE":"LAST_UPDATE","dates":"DATES"})

wkday = pd.DataFrame(wkday,columns =['NUMBER', 'CONTRACT_NAME', 'NAME', 'ADDRESS', 'POSITION','BANKING',
       'BONUS','BIKE_STANDS', 'AVAILABLE_BIKE_STANDS', 'AVAILABLE_BIKES',
       'STATUS', 'LAST_UPDATE', 'DATES'])
wkday

NUMBER CONTRACT_NAME                    NAME                 ADDRESS  \
0         42        dublin        SMITHFIELD NORTH        Smithfield North   
1         30        dublin    PARNELL SQUARE NORTH    Parnell Square North   
2         54        dublin          CLONMEL STREET          Clonmel Street   
3        108        dublin           AVONDALE ROAD           Avondale Road   
4         20        dublin       JAMES STREET EAST       James Street East   
...      ...           ...                     ...                     ...   
1125      39        dublin          WILTON TERRACE          Wilton Terrace   
1126      83        dublin              EMMET ROAD              Emmet Road   
1127      92        dublin  HEUSTON BRIDGE (NORTH)  Heuston Bridge (North)   
1128      21        dublin   LEINSTER STREET SOUTH   Leinster Street South   
1129      88        dublin         BLACKHALL PLACE         Blackhall Place   

                                               POSITION  BANKING  BONUS  \
0                  {'lat': 53.349562, 'lng': -6.278198}    False  False   
1     {'lat': 53.3537415547453, 'lng': -6.2653014478...    False  False   
2                   {'lat': 53.336021, 'lng': -6.26298}    False  False   
3                  {'lat': 53.359405, 'lng': -6.276142}    False  False   
4                  {'lat': 53.336597, 'lng': -6.248109}    False  False   
...                                                 ...      ...    ...   
1125               {'lat': 53.332383, 'lng': -6.252717}    False  False   
1126               {'lat': 53.340714, 'lng': -6.308191}    False  False   
1127               {'lat': 53.347802, 'lng': -6.292432}    False  False   
1128                {'lat': 53.34218, 'lng': -6.254485}    False  False   
1129                 {'lat': 53.3488, 'lng': -6.281637}    False  False   

      BIKE_STANDS  AVAILABLE_BIKE_STANDS  AVAILABLE_BIKES STATUS  \
0              30                     28                2   OPEN   
1              20                     18                2   OPEN   
2              33                     16               17   OPEN   
3              35                     22               13   OPEN   
4              30                     11               19   OPEN   
...           ...                    ...              ...    ...   
1125           20                     12                8   OPEN   
1126           40                     21               19   OPEN   
1127           40                      0               40   OPEN   
1128           30                      5               25   OPEN   
1129           30                     26                4   OPEN   

        LAST_UPDATE                DATES  
0     1670512585000  2022-12-08 15:16:25  
1     1670512436000  2022-12-08 15:13:56  
2     1670512607000  2022-12-08 15:16:47  
3     1670512550000  2022-12-08 15:15:50  
4     1670512642000  2022-12-08 15:17:22  
...             ...                  ...  
1125  1670523551000  2022-12-08 18:19:11  
1126  1670523583000  2022-12-08 18:19:43  
1127  1670523614000  2022-12-08 18:20:14  
1128  1670523629000  2022-12-08 18:20:29  
1129  1670523630000  2022-12-08 18:20:30  

[1130 rows x 13 columns]

In [5]:
wkend = pd.read_csv("wknd_dublin_bike.csv")
wkend = wkend.rename(columns = {"NUMBER": "NUMBER", "CONTRACT NAME":"CONTRACT_NAME", "name":"NAME", "address":"ADDRESS","position":"POSITION","BANKING":"BANKING",
       "bonus":"BONUS","latitude":"LATITUDE", "longitude":"LONGITUDE","BIKE STANDS":"BIKE_STANDS","AVAILABLE BIKE STANDS":"AVAILABLE_BIKE_STANDS","AVAILABLE BIKES":"AVAILABLE_BIKES",
       "status":"STATUS","LAST UPDATE":"LAST_UPDATE","dates":"DATES"})

wkend = pd.DataFrame(wkend,columns =['NUMBER', 'CONTRACT_NAME', 'NAME', 'ADDRESS', 'POSITION','BANKING',
       'BONUS','BIKE_STANDS', 'AVAILABLE_BIKE_STANDS', 'AVAILABLE_BIKES',
       'STATUS', 'LAST_UPDATE', 'DATES'])
wkend

NUMBER CONTRACT_NAME                    NAME                 ADDRESS  \
0         42        dublin        SMITHFIELD NORTH        Smithfield North   
1         30        dublin    PARNELL SQUARE NORTH    Parnell Square North   
2         54        dublin          CLONMEL STREET          Clonmel Street   
3        108        dublin           AVONDALE ROAD           Avondale Road   
4         20        dublin       JAMES STREET EAST       James Street East   
...      ...           ...                     ...                     ...   
1125      39        dublin          WILTON TERRACE          Wilton Terrace   
1126      83        dublin              EMMET ROAD              Emmet Road   
1127      92        dublin  HEUSTON BRIDGE (NORTH)  Heuston Bridge (North)   
1128      21        dublin   LEINSTER STREET SOUTH   Leinster Street South   
1129      88        dublin         BLACKHALL PLACE         Blackhall Place   

                                               POSITION  BANKING  BONUS  \
0                  {'lat': 53.349562, 'lng': -6.278198}    False  False   
1     {'lat': 53.3537415547453, 'lng': -6.2653014478...    False  False   
2                   {'lat': 53.336021, 'lng': -6.26298}    False  False   
3                  {'lat': 53.359405, 'lng': -6.276142}    False  False   
4                  {'lat': 53.336597, 'lng': -6.248109}    False  False   
...                                                 ...      ...    ...   
1125               {'lat': 53.332383, 'lng': -6.252717}    False  False   
1126               {'lat': 53.340714, 'lng': -6.308191}    False  False   
1127               {'lat': 53.347802, 'lng': -6.292432}    False  False   
1128                {'lat': 53.34218, 'lng': -6.254485}    False  False   
1129                 {'lat': 53.3488, 'lng': -6.281637}    False  False   

      BIKE_STANDS  AVAILABLE_BIKE_STANDS  AVAILABLE_BIKES STATUS  \
0              30                     24                6   OPEN   
1              20                     14                6   OPEN   
2              33                     26                7   OPEN   
3              35                     25               10   OPEN   
4              30                     27                3   OPEN   
...           ...                    ...              ...    ...   
1125           20                     20                0   OPEN   
1126           40                     21               19   OPEN   
1127           40                     18               22   OPEN   
1128           30                      8               22   OPEN   
1129           30                     23                7   OPEN   

        LAST_UPDATE                DATES  
0     1670671819000  2022-12-10 11:30:19  
1     1670672120000  2022-12-10 11:35:20  
2     1670672185000  2022-12-10 11:36:25  
3     1670671772000  2022-12-10 11:29:32  
4     1670672260000  2022-12-10 11:37:40  
...             ...                  ...  
1125  1670679966000  2022-12-10 13:46:06  
1126  1670679850000  2022-12-10 13:44:10  
1127  1670680245000  2022-12-10 13:50:45  
1128  1670680108000  2022-12-10 13:48:28  
1129  1670680288000  2022-12-10 13:51:28  

[1130 rows x 13 columns]

1. The average availabe bike on weekdays compared to weekends

In [42]:
wkday.AVAILABLE_BIKES.mean()

11.338053097345133

In [8]:
wkend.AVAILABLE_BIKES.mean()

12.172566371681416

This means that bike usage are more on weekdays compared to weekends because there are more available bikes on weekends i.e lesser bike stands

2. What is the busiest & quietest station during weekdays and weekends?
what address has the highest & lowest number of available bikes 

to answer question 2, we create a column for the percentage of bike stands by dividing the available bike stands by the number of bike stands

In [12]:
wkday["AVAILABLE_BIKE_STANDS%"]=wkday["AVAILABLE_BIKE_STANDS"]/wkday["BIKE_STANDS"]
wkend["AVAILABLE_BIKE_STANDS%"]=wkend["AVAILABLE_BIKE_STANDS"]/wkend["BIKE_STANDS"]

In [27]:
wk = []
df = wkday[['NAME','AVAILABLE_BIKE_STANDS%']].groupby(['NAME'])
df.sum().max()
value = df.sum().max()
key = df.sum().idxmax()
wk.append(key)
wk.append(value)
wk

[AVAILABLE_BIKE_STANDS%    MOUNTJOY SQUARE EAST
 dtype: object,
 AVAILABLE_BIKE_STANDS%    9.95
 dtype: float64]

MOUNTJOY SQUARE EAST becomes the busiest station with the most available bike stands percentage

In [28]:
wk = []
df = wkday[['NAME','AVAILABLE_BIKE_STANDS%']].groupby(['NAME'])
df.sum().max()
value = df.sum().min()
key = df.sum().idxmin()
wk.append(key)
wk.append(value)
wk

[AVAILABLE_BIKE_STANDS%    GEORGES QUAY
 dtype: object,
 AVAILABLE_BIKE_STANDS%    0.85
 dtype: float64]

Meanwhile, the quitest station on a weekday is GEORGES QUAY with just 0.85% of the avaialable bike stands pecentage 

For weekends

In [29]:
wk = []
df = wkend[['NAME','AVAILABLE_BIKE_STANDS%']].groupby(['NAME'])
df.sum().max()
value = df.sum().max()
key = df.sum().idxmax()
wk.append(key)
wk.append(value)
wk

[AVAILABLE_BIKE_STANDS%    GEORGES LANE
 dtype: object,
 AVAILABLE_BIKE_STANDS%    9.9
 dtype: float64]

Intrestingly, the busiest station on weekends appears to be Georges Lane with 9.9% of the available bike stands percentage

In [30]:
wk = []
df = wkend[['NAME','AVAILABLE_BIKE_STANDS%']].groupby(['NAME'])
df.sum().max()
value = df.sum().min()
key = df.sum().idxmin()
wk.append(key)
wk.append(value)
wk

[AVAILABLE_BIKE_STANDS%    FOWNES STREET UPPER
 dtype: object,
 AVAILABLE_BIKE_STANDS%    0.0
 dtype: float64]

Meanwhile, the quietest station is Fownes Street Upper

3. what is the number of available bikes for the answers to question 2 above?

In [105]:
wkday[wkday.NAME=="MOUNTJOY SQUARE EAST"].AVAILABLE_BIKES.sum()

2

The number of available bikes for MOUNTJOY SQUARE EAST is shown as 2, This further explains why it is the busiest station due to the low number of bikes available.

In [106]:
wkday[wkday.NAME=="GEORGES QUAY"].AVAILABLE_BIKES.sum()

183

Unlike the busiest station, Georges Quay has a total of 183 available bikes 

for weekends

In [107]:
wkend[wkend.NAME=="GEORGES LANE"].AVAILABLE_BIKES.sum()

4

The number of available bikes for the busiest station on weekends is 4

In [17]:
wkend[wkend.NAME=="FOWNES STREET UPPER"].AVAILABLE_BIKES.sum()

300

Fownes Street Upper has available bikes 300

4. The next question is to know if the presence of a credit card enabled machine would influence bike useage in both cases

In [40]:
wkday.groupby("BANKING").AVAILABLE_BIKES.agg(["mean"])

mean
BANKING           
False    11.203704
True     14.240000

In [41]:
wkend.groupby("BANKING").AVAILABLE_BIKES.agg(["mean"])

mean
BANKING           
False    12.208333
True     11.400000

In the presence of a credit card enabled machine on weekends, the available bikes are lesser compared to when its not, while on weekdays, the presence of credit card enabled machine promotes more available bikes comparred to when there's not.
which means that the presence of a credit card enable machine i.e payment terminal directly influences bike usage 

5. Given that the presence of a credit card enabled machine would influece the bike usage, our next question was to determine the total available bikes, bike stand and available bikes which has a credit card enabled machine in its facility during weekends and weekdays

In [18]:
grouped = wkday.groupby("BANKING")
grouped["AVAILABLE_BIKES", "AVAILABLE_BIKE_STANDS","BIKE_STANDS"].sum()

AVAILABLE_BIKES  AVAILABLE_BIKE_STANDS  BIKE_STANDS
BANKING                                                     
False              12100                  22726        34870
True                 712                    577         1290

In [135]:
grouped = wkend.groupby("BANKING")
grouped["AVAILABLE_BIKES", "AVAILABLE_BIKE_STANDS","BIKE_STANDS"].sum()

AVAILABLE_BIKES  AVAILABLE_BIKE_STANDS  BIKE_STANDS
BANKING                                                     
False              13185                  21672        34870
True                 570                    720         1290

The results indicate that there are more bike stands available on weekends when a credit card-enabled machine is available than there are on weekdays when there are less bike stands available. The station is operational since there are more available bike stands than there are available bikes.

The following results are reached after doing an exploratory analysis on Dublin bike usage on weekdays and weekends using our real-time dataset from JCDecaux;
1. More people use bicycles throughout the week than during the weekend.
2. MountJoy Square East, the busiest stop on weekdays, has less available bikes than the busiest station on weekends.
3. The availability of a credit card reader affects bike use on both weekdays and weekends.
In summary, people tend to utilize Dublin Bike more frequently during the weekdays when a payment terminal is present than during the weekends.